src/main/webapp/app/shared/components/code-button/code-button.component.ts

```diff
@@ -120,14 +123,40 @@ export class CodeButtonComponent implements OnInit, OnChanges {
         if (this.useSsh && this.sshEnabled && this.sshTemplateUrl) {
             return this.getSshCloneUrl(this.getRepositoryUri()) || this.getRepositoryUri();
         }
-
         if (this.isTeamParticipation) {
             return this.addCredentialsToHttpUrl(this.repositoryUriForTeam(this.getRepositoryUri()), insertPlaceholder);
         }
 
         return this.addCredentialsToHttpUrl(this.getRepositoryUri(), insertPlaceholder);
     }
 
+    loadVcsAccessToken() {
+        if (this.localVCEnabled && this.activeParticipation?.id && this.user) {
+            this.accountService.getVcsAccessToken(this.activeParticipation.id).subscribe({
+                next: (res: HttpResponse<string>) => {
+                    if (res.body) {
+                        this.user.vcsAccessToken = res.body;
```

Comment:
Since the token is exercise specific now, but it is stored once inside the user:

What happens when the user opens a different exercise after the current one?
I think that until the new specific access token could be loaded, the UI would show the wrong token (i.e. a broken clone URL).

Reply:
I guess the combination of lazy loading/generation and task specific tokens for users has a potential to break things or at least makes the implementation more complicated.

One idea for the PR:
As far as I've understood the whole idea of task specific tokens originate from the fact that the rotation of tokens could happen tight before a deadline of a task. A solution for that could be the modelling of exactly two tokens.

a) If a user regenerates a token on their behalf, clear all tokens and set the primary token.
b) If the end of life of a token is near (what near means should be configurable, because it depends on typical tasks): Primary token gets transferred to secondary token (but stays valid), a new primary token is generated

Artemis should always provide the primary token but also accept the secondary token until its eol.

I think that would ease the implementation because you only have to manage 2 Tokens. And it would also combine all benefits and erases the disadvantages of only one token.

What do you think about that idea?

Reply:
@b-fein  When a different exercise is opened, the ngOnInit+onChage load the user anew, which resets the `vcsAccessToken`. 

Four cases:
1. User does not have a user vcs token, fetching participation token fails: simply no token is displayed in the copy dialog.
2. User does have a user vcs token, fetching participation token succeeds: the participation token is used in the copy dialog.
3. User does have a user vcs token, fetching participation token fails, the user token is displayed in the dialog (the one currently used for gitlab setups), which also works for cloning.
4. User does not have a user vcs token, but fetching,  participation token exists: (current base case for localvc setup, until we add user token management like in gitlab), here the participation token is shown in the dialog.
So there isn't a race condition, in all cases (except when the server breaks down and doesn't provide any tokens anymore) the user obtains a token for authentication.

Reply:
@dfuchss Interesting idea, but I feel like that might be even more complicated as what we have right now

I think the current approach is as straightforward as it gets:
1. fetch token -> exists (base-case; tokens are still created when a user starts an exercise)
2. fetch token -> does not exist -> create new one (for participations created before this PR, or when tokens got deleted (for whatever reason))

Also it doesn't interfere with how the gitlab setup works currently

Reply:
Fine for me :)

What token do you plan the TAs and Instructors to use? I'm currently not getting this part :) Because TAs and esp. Instructors may have other permissions :)

Reply:
@SimonEntholzer can you please update the PR description to reflect all those ideas, especially this:
> 1. fetch token -> exists (base-case; tokens are still created when a user starts an exercise)
> 2. fetch token -> does not exist -> create new one (for participations created before this PR, or when tokens got deleted (for whatever reason))

Also: When is a token created (you removed the Migration that would do that for all old participations, right?)?